x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharFormatter.java

```diff
@@ -0,0 +1,371 @@
+        // also index the lower case version of the patterns if accepted
+        for (ToCharFormatter formatter : formatters) {
+            if (formatter.lowercaseAccepted) {
+                formatterMap.putIfAbsent(formatter.pattern.toLowerCase(Locale.ROOT), formatter);
+            }
+        }
+        FORMATTER_MAP = formatterMap;
+    }
+
+    @SuppressWarnings("OptionalGetWithoutIsPresent") private static final int
+        MAX_TO_CHAR_FORMAT_STRING_LENGTH =
+        FORMATTER_MAP.keySet().stream().mapToInt(String::length).max().getAsInt();
+
+    private final String pattern;
+    private final boolean lowercaseAccepted;
+    private final Function<String, String> fillMode;
+    private final boolean hasOrdinalSuffix;
+    private final Function<TemporalAccessor, String> formatFn;
+
+    private ToCharFormatter(
+        String pattern,
+        boolean lowercaseAccepted,
+        Function<String, String> fillMode,
+        boolean hasOrdinalSuffix,
+        Function<TemporalAccessor, String> formatFn) {
+
+        this.pattern = pattern;
+        this.lowercaseAccepted = lowercaseAccepted;
+        this.fillMode = fillMode;
+        this.hasOrdinalSuffix = hasOrdinalSuffix;
+        this.formatFn = formatFn;
+    }
+
+    private static Builder of(String pattern) {
+        return new Builder(pattern);
+    }
+
+    private String format(TemporalAccessor temporalAccessor) {
+        return formatFn.apply(temporalAccessor);
+    }
+
+    private ToCharFormatter withModifier(Function<String, String> modifierFn) {
+        return new ToCharFormatter(pattern, lowercaseAccepted, fillMode, hasOrdinalSuffix, formatFn.andThen(modifierFn));
+    }
+
+    private static List<ToCharFormatter> parsePattern(String toCharPattern) {
+        LinkedList<ToCharFormatter> formatters = new LinkedList<>();
+
+        boolean fillModeModifierActive = false;
+
+        while (toCharPattern.isEmpty() == false) {
+            boolean foundPattern = false;
+            for (int length = Math.min(MAX_TO_CHAR_FORMAT_STRING_LENGTH, toCharPattern.length()); length >= 1; length--) {
+                final String potentialPattern = toCharPattern.substring(0, length);
+                if ("FM".equals(potentialPattern) || "fm".equals(potentialPattern)) {
+                    // try to apply the fill mode modifier to the next formatter
+                    fillModeModifierActive = true;
+                    foundPattern = true;
+                } else if ("TH".equals(potentialPattern) || "th".equals(potentialPattern)) {
+                    // try to apply the ordinal suffix modifier to the last formatter
+                    ToCharFormatter lastFormatter = formatters.removeLast();
+                    if (lastFormatter == null) {
+                        lastFormatter = literal(potentialPattern);
+                    } else if (lastFormatter.hasOrdinalSuffix) {
+                        // if there was no previous pattern, simply prints the ordinal suffix pattern
```

Comment:
Should this comment go two lines up?

Then interestingly, for a `FM dd` pattern, the `FM` is just swallowed. Comparatively, the `dd TH` does print the `TH`. Not sure if so on purpose or not, but I guess it's acceptable, since Postgres behaves exactly the same.

Reply:
The behaviour currently matches exactly the Postgres behaviour. The patterns are tested both with and without the `FM` and `TH` modifiers.